### ISSUE TITLE
Bash completion broken after running `nix develop`

### ISSUE THREAD

**Issue by x10an14** [NixOS/nix#6091]
**Describe the bug**

Two repoes to try this out with (flaked projects):
- https://git.sr.ht/~x10an14/pass-steam
- https://github.com/kamadorueda/alejandra

**Steps To Reproduce**
1. `git clone` a flaked repo.
2. Type in `git st<tab><tab>` and see `status` show up as an auto completion.
3. Invoke `nix develop`.
4. Repeat step 2 and see that the tab completion does not work anymore.

**Expected behavior**

`nix develop` not to break tab completion in bash.

**`nix-env --version` output**
```
nix-env (Nix) 2.6.0
```

**Additional context**

Experienced by both me and @tomberek on each our machines/configs in each our repos, ref. [this nix office hour](https://discourse.nixos.org/t/general-nix-office-hours/15019/35?u=x10an14).

**Comment by NobbZ** [comment]
This is by design and was already the same back with `nix shell`(2.3)/`nix-shell`.

You are kicked into `stdenv`s bash to recreate an authentic environment as used by the nix build. There are a couple of workarounds:

1. use `direnv` with `nix-direnv`s `use flake`
2. use `direnv` with a `use_flake` function as described in the wiki (https://nixos.wiki/wiki/Flakes#Direnv_integration)
3. use `nix shell nixpkgs#bashInteractive -c bash` to get into a subshell first which you can leave to clean up the environment, then use `eval "$(nix print-dev-env)"` to actually load the dev env into your current shell.

Each of these does work with ZSH as well. I doubt that 3. can be made with fish.

**Comment by x10an14** [comment]
Thanks for a quick answer! Nice to know it's by design! Unfortunately I got none of your three suggestions to work for me, using the aforementioned "Steps to reproduce" above =/

(I additionally deleted the `.direnv` folder cache, toggled `direnv allow/deny`, and spawned a new shell for each test - `git st<tab><tab>` always worked up to the point of invoking `nix develop`/`eval "$(nix print-dev-env)"`.)

### My attempts at your suggest workarounds
> (...) There are a couple of workarounds:
> 1. use `direnv` with `nix-direnv`s `use flake`

I thought I already had this, but I realize now that I don't have the `nix-direnv` package installed (just the `direnv` one).
Where would I as a end-user find documentation that this is an option? (Maybe even the preferred/recommended one - at least to start with)?

I'm gonna skip this option for now though, since it hard-codes the `nix` version to 2.4... =(

---
> 2. use `direnv` with a `use_flake` function as described in the wiki (https://nixos.wiki/wiki/Flakes#Direnv_integration)

I did take a look at that page when I began my foray into Flakes, I'm sorry to say it was out of my competency to grok it then - you helped me figure it out now! =)

Still can't get it to work though =( Here's my `.envrc` diff in my referenced project up above when I tested it just now:
```
diff --git a/.envrc b/.envrc
index 3550a30..f74489a 100644
--- a/.envrc
+++ b/.envrc
@@ -1 +1,7 @@
-use flake
+use_flake() {
+  watch_file flake.nix
+  watch_file flake.lock
+  eval "$(nix print-dev-env --profile "$(direnv_layout_dir)/flake-profile")"
+}
+
+use_flake
```
---
> 3. use `nix shell nixpkgs#bashInteractive -c bash` to get into a subshell first which you can leave to clean up the environment, then use `eval "$(nix print-dev-env)"` to actually load the dev env into your current shell.

I actually don't get this to work either.

**Comment by tomberek** [comment]
In order of "purity":
```
nix build
```
```
nix develop -i
```
```
nix develop
```
```
nix shell nixpkgs#bashInteractive --command bash
. <(nix print-dev-env)
```
```
. <(nix print-dev-env)
```
The annoyance of the two commands can be simplified via some other tools like direnv, but I propose it is common enough to need an easy way to utilize without third-party tooling.

**Comment by jonringer** [comment]
related: https://github.com/NixOS/nix/issues/5131

**Comment by tomberek** [comment]
> However, https://github.com/NixOS/nix/issues/6091#issue-1133997448 seems to be saying that both you and @x10an14 are seeing nix develop put you in a shell using bash (not bashInteractive).

Not quite, it's slightly more convoluted. The shell spawned is bash-interactive (readlink /proc/$$/exe) , but the SHELL variable and PATH refer to the non-interactive shell. This is what causes starting more subshells to turn into this annoying behavior:
```
[tom@tframe:~/project]$ nix develop
[tom@tframe:~/project]$ $SHELL
bash: shopt: dirspell: invalid shell option name
bash: shopt: hostcomplete: invalid shell option name
complete: command not found
bash: set: vi: invalid option name
bind: command not found
bind: command not found
bind: command not found
bind: command not found
bind: command not found
bind: command not found
bind: command not found
\[\]
\[\][\[\]tom@tframe:~/project]$\[\] exit
```

The part that affects bash completions is that the environment is cleared out and then bash-interactive is called with `--rcfile` which bypasses some of the profile/bash_profile/bashrc setup.

Part of this is a design question about what the intention and user experience should be. I fear we are running into a similar expansion of usage of `nix develop` that happened with `nix-shell`: usage for both debugging a build in a sandbox-identical environment as well as developer environments. Perhaps the `nix run/shell/develop` triple is not enough and a `nix run/shell/develop/debug` is needed? That comes at a cost of another concept and subcommand to confuse people. Another option is to allow `nix develop` to be less pure and point people to `nix develop -i` in those cases where it is a problem.

**Comment by tomberek** [comment]
@x10an14 you seem to want this (or the equivalent for your $SHELL):
```
. <(nix print-dev-env)
```

**Comment by ncfavier** [comment]
For me the main problem seems to be that `nix develop/print-dev-env` overwrites `XDG_DATA_DIRS` so that loadable completion isn't loaded anymore.

Maybe nix should maintain a list of variables to prepend to rather than overwrite, [as it does for `PATH`](https://github.com/NixOS/nix/blob/bbe7187db99ee46691e87e8eb8baded850cc3035/src/nix/develop.cc#L283-L287)?

**Comment by ncfavier** [comment]
https://github.com/NixOS/nix/pull/6702

**Comment by jessestricker** [comment]
would be fixed by #7010
